logstage/logstage-sink-file/src/main/scala/com/github/pshirshov/izumi/FileSink.scala

```diff
@@ -0,0 +1,88 @@
+package com.github.pshirshov.izumi
+
+import java.util.concurrent.atomic.AtomicReference
+import com.github.pshirshov.izumi.models._
+import com.github.pshirshov.izumi.logstage.api.logger.RenderingPolicy
+import com.github.pshirshov.izumi.logstage.model.Log
+import com.github.pshirshov.izumi.logstage.model.logger.LogSink
+import com.github.pshirshov.izumi.logstage.sink.console.ConsoleSink
+import scala.util.{Failure, Success, Try}
+
+
+class FileSink(policy: RenderingPolicy, consoleSink: ConsoleSink, config: FileSinkConfig) extends LogSink {
+
+  private val sinkState = new AtomicReference[SinkState](SinkState(config = config))
+
+  def processCurrentFile(state: SinkState): Try[SinkState] = {
+    val curFileId = state.currentFileId
+    val updatedStatus = {
+      if (state.currentFileSize == FileSize.undefined) {
+        state.getFile(curFileId).flatMap {
+          file =>
+            file.getFileSize.map {
+              case size if size > 0 =>
+                state.config.rotation.filesLimit.map(l => {
+                  (curFileId + 1) to l map {
+                    id => state.removeFile(id)
+                  }
+                })
+                state.copy(currentFileSize = size)
+              case size =>
+                state.copy(currentFileSize = size)
+            }
+        }
+      } else {
+        Success(state)
+      }
+    }
+
+    updatedStatus.flatMap {
+      st =>
+        Success {
+          st.currentFileSize match {
+            case size if size == st.getMaxSize =>
+              st.copy(currentFileId = curFileId + 1, currentFileSize = FileSize.init)
+            case _ =>
+              st
+          }
+        }
+    }
+  }
+
+  def adjustByRotation(state: SinkState): Try[SinkState] = {
+    state.config.rotation.performRotate(state)
+  }
+
+  def writeLogEntry(oldState: SinkState, payload: String): Try[SinkState] = {
```

Comment:
I guess if you get rid of `Try` in favor of an abstract container your effect separation would be better.

Reply:
You can't have an abstract container without importing cats or reinventing part of it (you don't have flatMap).
Instead, you can just not use any monad at all, as in: https://github.com/kaishh/izumi-r2/commit/f34816baf45b94a0d08d2f48597aee3d826889c4#diff-cf45bad449f68fde48a48e44a4ffb79b